### PR TITLE
[inductor] Improved indexing simplification and loop body representation

### PR DESCRIPTION
**Pull request by jansel**

### CODE REVIEW

File: torchinductor/sizevars.py [hunk 1]
```diff
@@ -207,3 +280,67 @@ def codegen_shape_tuple(self, shape):
         if len(parts) == 1:
             return f"({parts[0]}, )"
         return f"({', '.join(parts)})"
+
+
+def join_dimensions(expr: sympy.Expr) -> sympy.Expr:
+    """
+    ModularIndexing(i0, 1, 32) + 32 * ModularIndexing(i0, 32, 4)
```
Comment (ngimel) [inline torchinductor/sizevars.py:287]: wow, that's pretty cool!

File: tests/test_torchinductor.py [hunk 2]
```diff
@@ -165,6 +169,84 @@ class SweepInputsCpuTest(SweepInputs2, TestCase):
 SweepInputsCpuTest.populate()
 
 
+class TestIndexingSimplification(unittest.TestCase):
+    def test_indexing_simplification(self):
```
Comment (ngimel) [inline tests/test_torchinductor.py:173]: tests are pretty cool and show how simplifications are supposed to work, but (if we are to use them as documentation) would benefit from brief comments on why those results are expected)

File: torchinductor/codegen/triton.py [hunk 3]
```diff
@@ -279,8 +284,12 @@ def indexing(self, index: sympy.Expr):
         offset = index.subs({v: 0 for v in chain(iter_vars, reduction_vars)})
         base_part = index.subs({v: 0 for v in reduction_vars}) - offset
         reduction_part = index.subs({v: 0 for v in iter_vars}) - offset
-
-        assert index == offset + base_part + reduction_part
+        assert index == offset + base_part + reduction_part, (
```
[the remark this answers is not in the document]
Comment (jansel) [reply torchinductor/codegen/triton.py]: fixed

File: tests/test_torchinductor.py (hunk 2 above)
Comment (jansel) [reply tests/test_torchinductor.py:173]: fixed

**Review by ngimel — APPROVED** [review]
That's very cool

**Pull request merged**